Split this layer into meta-chromium and meta-firefox

shr-project opened the pull request:
* they don't share the same layer dependencies, no bbclasses, even the people who maintain them are different and firefox isn't as well maintained as chromium

* splitting them allows us to drop meta-python2 (soon) and optional meta-rust dependency from meta-chromium
  and meta-clang from meta-firefox

shr-project commented:
@otavio @rakuco @kraj @msisov @ashie @fbertux

It's just a draft, because I haven't finished testing it, but I would like to hear your thoughts about this.

shr-project commented:
Looks like python2-less build still doesn't work
```
python ../../third_party/blink/renderer/bindings/scripts/generate_bindings.py --web_idl_database gen/third_party/blink/renderer/bindings/web_idl_database.pickle --root_src_dir ../../ --root_gen_dir gen --output_core_reldir third_party/blink/renderer/bindings/core/v8/ --output_modules_reldir third_party/blink/renderer/bindings/modules/v8/ enumeration callback_function callback_interface interface namespace
/usr/bin/env: /usr/bin/env: ‘python2’: No such file or directory
‘python2’: No such file or directory
/usr/bin/env: ‘python2’: No such file or directory
```

will update to keep meta-python2 dependency in meta-chromium.

rakuco commented on meta-firefox/README:
```diff
@@ -0,0 +1,48 @@
+list openembedded-devel@lists.openembedded.org (you have to be subscribed to
+post to the list). Please cc the maintainers if you send your patches.
+
+Maintainers: Fabio Berton <fabio.berton@ossystems.com.br>
```
I think this list needs to be updated; Maxim and I don't work on Firefox, for example.

shr-project replied:
That's definitely open for discussion, I was wondering if you want to have one README about submitting patches in meta-browser and then README in meta-chromium/meta-firefox documenting the required layer dependencies and different set of maintainer (is anyone maintaining firefox? I was the last one to update it, just because I needed to fix build with newer rust for benchmarking build time, but I have no interest in maintaining it and I don't know how much @ashie is still interested in it), maybe the README file there should say that we're looking for (new/more) maintainers there.

And with separate READMES in meta-chromium/meta-firefox it might make sense to merge them with ./meta-chromium/recipes-browser/chromium/README ./meta-firefox/recipes-browser/firefox/README.md as well.

rakuco reviewed (commented):
I don't know if all the magical incantations being done to the Firefox files make sense, but this sounds like an interesting idea to me.!

otavio commented:
I think this is indeed a good change. I see no problem in doing this.

kraj commented on meta-firefox/conf/layer.conf:
```diff
@@ -0,0 +1,115 @@
+   firefox-l10n-zh-tw \
+ "
+
+BBFILE_COLLECTIONS += "browser-layer"
```
this should be called something else perhaps `mozilla-layer` or some such

shr-project replied:
it is called "firefox-browser-layer" if you look at 2nd commit.

I've intentionally kept 1st commit as just move + copy of README + layer.conf and did the modifications in 2nd so it's easier to review.

kraj reviewed (commented):
I think firefox also needs clang now a days, although it works with gcc but their main builds are using clang. So lets keep those defaults. see https://glandium.org/blog/?p=3888

kraj reviewed (commented):
I think we should still use clang to compile, but if upstream is still using gcc for default builds then we could keep using gcc

shr-project commented:
> I think firefox also needs clang now a days, although it works with gcc but their main builds are using clang. So lets keep those defaults. see https://glandium.org/blog/?p=3888

That is indeed true for nowadays, but this is firefox 68 from 2019-07-09 and I think the recipe needs some changes to actually use clang (even when meta-clang is included).

As it is now it doesn't compile with latest rust 1.49.0 from meta-rust and the last rust version this firefox is compatible with (1.39.0 after https://github.com/OSSystems/meta-browser/pull/378) now isn't in meta-rust anymore:
https://github.com/meta-rust/meta-rust/issues/298#issuecomment-775137561

So the motivation behind this split was also to separate less-maintained firefox from well-maintained chromium.

shr-project commented:
> Looks like python2-less build still doesn't work
> 
> ```
> python ../../third_party/blink/renderer/bindings/scripts/generate_bindings.py --web_idl_database gen/third_party/blink/renderer/bindings/web_idl_database.pickle --root_src_dir ../../ --root_gen_dir gen --output_core_reldir third_party/blink/renderer/bindings/core/v8/ --output_modules_reldir third_party/blink/renderer/bindings/modules/v8/ enumeration callback_function callback_interface interface namespace
> /usr/bin/env: /usr/bin/env: ‘python2’: No such file or directory
> ‘python2’: No such file or directory
> /usr/bin/env: ‘python2’: No such file or directory
> ```
> 
> will update to keep meta-python2 dependency in meta-chromium.

And there is unfortunately also dependency on target python from setuptools inherit, which doesn't play well with:
https://lists.openembedded.org/g/openembedded-devel/message/90046

so I'm trying to get rid of that now (just removing setuptools inherit doesn't work as well (fails with "| ImportError: No module named pkg_resources"), but I think a dependency on native package should be enough, will return to that after more build tests of firefox.

shr-project commented:
> > I think firefox also needs clang now a days, although it works with gcc but their main builds are using clang. So lets keep those defaults. see https://glandium.org/blog/?p=3888
> 
> That is indeed true for nowadays, but this is firefox 68 from 2019-07-09 and I think the recipe needs some changes to actually use clang (even when meta-clang is included).
> 
> As it is now it doesn't compile with latest rust 1.49.0 from meta-rust and the last rust version this firefox is compatible with (1.39.0 after #378) now isn't in meta-rust anymore:
> [meta-rust/meta-rust#298 (comment)](https://github.com/meta-rust/meta-rust/issues/298#issuecomment-775137561)
> 
> So the motivation behind this split was also to separate less-maintained firefox from well-maintained chromium.

nevermind, I've tried without meta-clang and it fails now with
```
|  0:03.56 checking rustc version... 1.49.0
|  0:03.57 checking cargo version... 1.49.0
|  0:03.59 checking rust host... x86_64-linux
|  0:03.59 checking rust target... x86_64-oe-linux
|  0:03.59 checking for rust target triplet... x86_64-oe-linux
|  0:03.59 checking for rust host triplet... x86_64-linux
|  0:03.59 checking for rustdoc... /OE/build/oe-core/tmp-glibc/work/core2-64-oe-linux/firefox/68.9.0esr-r0/recipe-sysroot-native/usr/bin/rustdoc
|  0:03.59 checking for cbindgen... /OE/build/oe-core/tmp-glibc/work/core2-64-oe-linux/firefox/68.9.0esr-r0/recipe-sysroot-native/usr/bin/cbindgen
|  0:03.59 checking for rustfmt... not found
|  0:03.60 checking for llvm-config... not found
|  0:03.60 ERROR: Could not find LLVM/Clang installation for compiling bindgen.
|  0:03.60 Please specify the 'LLVM_CONFIG' environment variable
|  0:03.60 (recommended), pass the '--with-libclang-path' and '--with-clang-path'
|  0:03.60 options to configure, or put 'llvm-config' in your PATH.  Altering your
|  0:03.60 PATH may expose 'clang' as well, potentially altering your compiler,
|  0:03.60 which may not be what you intended.
|  0:03.63 *** Fix above errors and then restart with\
|  0:03.63                "./mach build"
|  0:03.63 make: *** [client.mk:115: configure] Error 1
| WARNING: exit code 2 from a shell command.
|
NOTE: recipe firefox-68.9.0esr-r0: task do_configure: Failed
```

so something changed since last time I was building this, will leave meta-clang dependency in meta-firefox (build running now).

shr-project commented:
In last version I've:
* Merged the READMEs
* restored meta-clang dependency in meta-firefox
* Removed @rakuco and @msisov from meta-firefox.
* Added a fix for firefox with latest meta-rust.
* Removed target python dependency from chromium.

Firefox already built OK, testing chromium-x11 now.

ashie commented:
Sorry for my less activity.

> So the motivation behind this split was also to separate less-maintained fire\
fox from well-maintained chromium.

I also think they should be separated, it's a pretty good modification.
As for putting them in the same place, there are more demerits than merits.

> I don't know how much @ashie is still interested in it

I'm still interesting it but I have no time to maintain it in actual.
I may upgrade it when I have time again, but I cannot guarantee it for now.

shr-project commented:
Rebased on latest master (with chromium upgrade merged) and added support for hardknott in LAYERSERIES_COMPAT.

Please review and merge :).

otavio commented:
@shr-project it is conflicting. It'd be nice if you could rebase it one more time.

If people agree, I think we ought to merge this and do whatever rework as a new PR as this will keep conflicting. I think this is a step in the right direction as well.

shr-project commented:
> @shr-project it is conflicting. It'd be nice if you could rebase it one more time.

I did on latest master, but pushed to different remote than this PR is using, fixed now, thanks.

rakuco commented on meta-chromium/README.md:
```diff
@@ -1,5 +1,56 @@
+* Khem Raj <raj.khem@gmail.com>
+* Otavio Salvador <otavio@ossystems.com.br>
+* Maksim Sisov <msisov@igalia.com>
+* Takuro Ashie <ashie@clear-code.com>
```
I think @ashie can be removed from this list.

shr-project replied:
@ashie can you confirm?

ashie replied:
Yeah, please remove it.

shr-project replied:
Done in https://github.com/OSSystems/meta-browser/pull/481

rakuco commented on meta-chromium/conf/layer.conf:
```diff
@@ -0,0 +1,14 @@
+BBFILE_PRIORITY_chromium-browser-layer = "7"
+
+LAYERVERSION_chromium-browser-layer = "1"
+LAYERSERIES_COMPAT_chromium-browser-layer = "dunfell gatesgarth hardknott"
```
I'd rather land the commit adding _hardknott_ to `LAYERSERIES_COMPAT` separately, as it's unrelated to the rest of the PR.

shr-project replied:
It's needed to build these layers with latest master now, so I can split it to separate PR, but both should be applied now.

rakuco replied:
That's OK, I'll gladly approve the other PR, I just don't like the idea of landing unrelated commits together as part of a single PR since it makes digging through this repository's history more complicated.

rakuco commented on meta-firefox/recipes-browser/firefox/firefox_68.9.0esr.bb:
```diff
@@ -25,6 +25,7 @@ SRC_URI = "https://ftp.mozilla.org/pub/firefox/releases/${PV}/source/firefox-${P
            file://fixes/Bug-1556197-amend-Bug-1544631-for-fixing-mips32.patch \
            file://fixes/Bug-1560340-Only-add-confvars.sh-as-a-dependency-to-.patch \
            file://fixes/bug1545437-enable-to-specify-rust-target.patch \
+           file://fixes/Bug-1640982-SetCARGO_PROFILE_RELEASE_LTO-true-when-e.patch \
```
Isn't this unrelated to this PR too?

shr-project replied:
Well, it's not needed for the split, but it's needed to at least build test firefox with current meta-rust and I wanted to build it after split, just to confirm I didn't break it.

rakuco replied:
Same as above: I'm happy you've done these tests; now that we're sure everything works this commit can be landed separately once this PR is merged.

rakuco reviewed (commented):
I'm glad to see this progressing! Just to confirm something that looks obvious: all users will need to update their `bblayers.conf` and point to `meta-browser/meta-{chromium,firefox}`, right?

shr-project commented:
> I'm glad to see this progressing! Just to confirm something that looks obvious: all users will need to update their `bblayers.conf` and point to `meta-browser/meta-{chromium,firefox}`, right?

Correct.

shr-project commented:
@rakuco 
https://github.com/OSSystems/meta-browser/pull/476
https://github.com/OSSystems/meta-browser/pull/474
https://github.com/OSSystems/meta-browser/pull/475

rakuco reviewed (approved):
lgtm, thank you. Please just address my comment in meta-chromium/README.md before landing.

otavio commented:
@rakuco I am sorry, I read your comment after merging.

otavio commented:
@shr-project the other PRs will need rebase. Can you take care of it?

shr-project commented:
> @shr-project the other PRs will need rebase. Can you take care of it?

done

shr-project commented:
I've also rebased @kraj 's https://github.com/OSSystems/meta-browser/pull/465 and https://github.com/OSSystems/meta-browser/pull/466 on top of the layer split, but he probably didn't click the "Allow edits by maintainers" on PR, but the rebase was relatively straight forward, so he won't have any issues doing it locally if he prefers (sent him the same over gtalk).

denix0 commented:
> Looks like python2-less build still doesn't work
> 
> will update to keep meta-python2 dependency in meta-chromium.

See https://bugs.chromium.org/p/chromium/issues/detail?id=942720 for latest updates on the Python3 migration. There are still few blockers remaining...

shr-project commented:
> > Looks like python2-less build still doesn't work
> > will update to keep meta-python2 dependency in meta-chromium.
> 
> I'm sure you already know this, but see https://bugs.chromium.org/p/chromium/issues/detail?id=942720 for latest updates on the Python3 migration. There are still few blockers remaining...
> Anyway, thanks for making this split happen!

Yes, I was just curious if the last update there from @rakuco is enough to get the parts we build here compatible already or how much is missing for meta-chromium build.

but his change https://chromium.googlesource.com/chromium/src/+/9083c75ed2388be67d86a04521a78f9230d70b53 isn't included in 89.0.4389.90 yet, so we either need to wait for newer version or backport this and try if it helps.

Maybe @rakuco can share some status details as he was working on it recently.

rakuco commented:
The move to Python 3 is still in progress upstream. See the other bugs that that bug is blocked on, for example, or https://bugs.chromium.org/p/chromium/issues/detail?id=1112471 more specifically. Given how many commits are still landing everywhere in the tree to get it working, the M89 in our recipe is not ready for Python 3 yet. I'd wait for an official announcement upstream switching to Python 3 (hopefully during the M91-93 time frame) before considering dropping meta-python2 here.
